# CSI5155 Machine Learning Assignment 1

## Installing Prerequisites and Dependencies

In [1]:
%pip install pandas
%pip install numpy
%pip install scikit-learn
%pip install joblib

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip available: 22.2.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd;
import numpy as np;
from models import Models;
import constants;
from sklearn.model_selection import train_test_split;
import random;
from fileOrganizer import organize;
import joblib;

## Load the Dataset

In [3]:
df_all = pd.read_csv('drug_consumption.data', sep=',', header=None, names=constants.column_names_orig);
df_all

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore  ...  ecstasy  heroin  ketamine legalh  lsd meth  \
0    -0.58331 -0.91699  ...      CL0     CL0       CL0    CL0  CL0  CL0   
1     1.43533  0.76096  ...      CL4     CL0       CL2    CL0  CL2  CL3   
2    -0.84732 -1.62090  ...      CL0     CL0       CL0    CL0  CL0  CL0   
3    -0.01928  0.59042  ...      CL0     CL0       CL2    CL0  CL0  CL0   
4    -0.45174 -0.30172  ...      CL1     CL0       CL0    CL1  CL0  CL0   
...       ...      ...  ...      ...     ...       ...    ...  ...  ...   
1880  1.88511  0.76096  ...      CL0     CL0       CL0    CL3  CL3  CL0   
1881  0.58331  0.76096  ...      CL2     CL0       CL0    CL3  CL5  CL4   
1882 -1.27553 -1.77200  ...      CL4     CL0       CL2    CL0  CL2  CL0   
1883  0.29338 -1.62090  ...      CL3     CL0       CL0    CL3  CL3  CL0   
1884  1.65653  1.11406  ...      CL3     CL0       CL0    CL3  CL3  CL0   

     mushrooms nicotine semer  vsa  
0          CL0      CL2   CL0  CL0  
1          CL0      CL4   CL0  CL0  
2          CL1      CL0   CL0  CL0  
3          CL0      CL2   CL0  CL0  
4          CL2      CL2   CL0  CL0  
...        ...      ...   ...  ...  
1880       CL0      CL0   CL0  CL5  
1881       CL4      CL5   CL0  CL0  
1882       CL2      CL6   CL0  CL0  
1883       CL3      CL4   CL0  CL0  
1884       CL3      CL6   CL0  CL2  

[1885 rows x 32 columns]

Observations:
- The original dataset consists of 1885 instances. 
- Each of which has 32 features in total. 

## Handling Data in the dataset to match with our problem

### Group and Convert multiple categories into 2 unique categories (for the binary classification)

In [4]:
num_types = ['int64', 'float64'];
for i in df_all.columns:
    if (df_all[i].dtype not in num_types):
        df_all[i] = df_all[i].apply(lambda x: "non-user" if int(x.replace("CL", "")) <= 1 else "user");
df_all

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore  ...   ecstasy    heroin  ketamine    legalh       lsd  \
0    -0.58331 -0.91699  ...  non-user  non-user  non-user  non-user  non-user   
1     1.43533  0.76096  ...      user  non-user      user  non-user      user   
2    -0.84732 -1.62090  ...  non-user  non-user  non-user  non-user  non-user   
3    -0.01928  0.59042  ...  non-user  non-user      user  non-user  non-user   
4    -0.45174 -0.30172  ...  non-user  non-user  non-user  non-user  non-user   
...       ...      ...  ...       ...       ...       ...       ...       ...   
1880  1.88511  0.76096  ...  non-user  non-user  non-user      user      user   
1881  0.58331  0.76096  ...      user  non-user  non-user      user      user   
1882 -1.27553 -1.77200  ...      user  non-user      user  non-user      user   
1883  0.29338 -1.62090  ...      user  non-user  non-user      user      user   
1884  1.65653  1.11406  ...      user  non-user  non-user      user      user   

          meth mushrooms  nicotine     semer       vsa  
0     non-user  non-user      user  non-user  non-user  
1         user  non-user      user  non-user  non-user  
2     non-user  non-user  non-user  non-user  non-user  
3     non-user  non-user      user  non-user  non-user  
4     non-user      user      user  non-user  non-user  
...        ...       ...       ...       ...       ...  
1880  non-user  non-user  non-user  non-user      user  
1881      user      user      user  non-user  non-user  
1882  non-user      user      user  non-user  non-user  
1883  non-user      user      user  non-user  non-user  
1884  non-user      user      user  non-user      user  

[1885 rows x 32 columns]

Observation:
- All categorical fields (which represents whether the person has taken that drug before) has been replaced with values suitable for  a binary classfication problem.

In [5]:
# save the modified DataFrame into a new .csv file
df_all.to_csv('drug_consumption.modified', index=False)

### Construct 2 datasets

#### People who consume Chocolate

In [6]:
# find a list of column names in consideration
cols = [];
for i in df_all.columns:
    if (df_all[i].dtype in num_types):
        cols.append(i);
cols.append(constants.choco_dataset);
cols

['ID',
 'age',
 'gender',
 'education',
 'country',
 'ethnicity',
 'nscore',
 'escore',
 'oscore',
 'ascore',
 'cscore',
 'impuslive',
 'ss',
 'choc']

In [7]:
df_dataset_choco = df_all[cols];
df_dataset_choco

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore   cscore  impuslive       ss  choc  
0    -0.58331 -0.91699 -0.00665   -0.21712 -1.18084  user  
1     1.43533  0.76096 -0.14277   -0.71126 -0.21575  user  
2    -0.84732 -1.62090 -1.01450   -1.37983  0.40148  user  
3    -0.01928  0.59042  0.58489   -1.37983 -1.18084  user  
4    -0.45174 -0.30172  1.30612   -0.21712 -0.21575  user  
...       ...      ...      ...        ...      ...   ...  
1880  1.88511  0.76096 -1.13788    0.88113  1.92173  user  
1881  0.58331  0.76096 -1.51840    0.88113  0.76540  user  
1882 -1.27553 -1.77200 -1.38502    0.52975 -0.52593  user  
1883  0.29338 -1.62090 -2.57309    1.29221  1.22470  user  
1884  1.65653  1.11406  0.41594    0.88113  1.22470  user  

[1885 rows x 14 columns]

Observations:
- We have 14 features in consideration in the first dataset.
- Meanwhile, the number of instances remains unchanged. 

#### People who consume Magic Mushrooms

In [8]:
# find a list of column names in consideration
cols = [];
for i in df_all.columns:
    if (df_all[i].dtype in num_types):
        cols.append(i);
cols.append(constants.mushrooms_dataset);
cols

['ID',
 'age',
 'gender',
 'education',
 'country',
 'ethnicity',
 'nscore',
 'escore',
 'oscore',
 'ascore',
 'cscore',
 'impuslive',
 'ss',
 'mushrooms']

In [9]:
df_dataset_mushrooms = df_all[cols];
df_dataset_mushrooms

ID      age   gender  education  country  ethnicity   nscore   escore  \
0        1  0.49788  0.48246   -0.05921  0.96082    0.12600  0.31287 -0.57545   
1        2 -0.07854 -0.48246    1.98437  0.96082   -0.31685 -0.67825  1.93886   
2        3  0.49788 -0.48246   -0.05921  0.96082   -0.31685 -0.46725  0.80523   
3        4 -0.95197  0.48246    1.16365  0.96082   -0.31685 -0.14882 -0.80615   
4        5  0.49788  0.48246    1.98437  0.96082   -0.31685  0.73545 -1.63340   
...    ...      ...      ...        ...      ...        ...      ...      ...   
1880  1884 -0.95197  0.48246   -0.61113 -0.57009   -0.31685 -1.19430  1.74091   
1881  1885 -0.95197 -0.48246   -0.61113 -0.57009   -0.31685 -0.24649  1.74091   
1882  1886 -0.07854  0.48246    0.45468 -0.57009   -0.31685  1.13281 -1.37639   
1883  1887 -0.95197  0.48246   -0.61113 -0.57009   -0.31685  0.91093 -1.92173   
1884  1888 -0.95197 -0.48246   -0.61113  0.21128   -0.31685 -0.46725  2.12700   

       oscore   ascore   cscore  impuslive       ss mushrooms  
0    -0.58331 -0.91699 -0.00665   -0.21712 -1.18084  non-user  
1     1.43533  0.76096 -0.14277   -0.71126 -0.21575  non-user  
2    -0.84732 -1.62090 -1.01450   -1.37983  0.40148  non-user  
3    -0.01928  0.59042  0.58489   -1.37983 -1.18084  non-user  
4    -0.45174 -0.30172  1.30612   -0.21712 -0.21575      user  
...       ...      ...      ...        ...      ...       ...  
1880  1.88511  0.76096 -1.13788    0.88113  1.92173  non-user  
1881  0.58331  0.76096 -1.51840    0.88113  0.76540      user  
1882 -1.27553 -1.77200 -1.38502    0.52975 -0.52593      user  
1883  0.29338 -1.62090 -2.57309    1.29221  1.22470      user  
1884  1.65653  1.11406  0.41594    0.88113  1.22470      user  

[1885 rows x 14 columns]

Observations:
- We have 14 features in consideration in the second dataset.
- Meanwhile, the number of instances also remains unchanged. 

## A. Machine Learning - Supervised Learning

### Classification task for Chocolate users

#### Load the Models

In [10]:
chocoModels = Models();
resChocoModels = chocoModels.getModels(
    constants.descisionTree, # Single Decision Tree classifier
    constants.randForest, # Random Forest classifier
    constants.svm, # Support Vector Machine (SVM) classifier - SVC with RBF
    constants.gradientBoost, # gradient boosting (GB) ensemble
    constants.mlp, # multi-layer perceptron (MLP) classifier
    constants.knn # k‐nearest neighbour (k-NN) classifier
);
print(resChocoModels);
# save all models
chocoModels.saveModels();

{'decision tree': DecisionTreeClassifier(criterion='entropy', splitter='random'), 'random forest': RandomForestClassifier(criterion='entropy'), 'SVM': SVC(), 'gradient boosting': GradientBoostingClassifier(), 'multi-layer perceptron (MLP)': MLPClassifier(hidden_layer_sizes=3769, learning_rate='adaptive'), 'k‐nearest neighbour (k-NN) classifier': KNeighborsClassifier(n_jobs=-1, n_neighbors=7, weights='distance')}


#### Split the Dataset

In [11]:
# Convert the dataset to a numpy array
np_samples_choco = df_dataset_choco.iloc[:, :-1].to_numpy();
np_labels_choco = df_dataset_choco.iloc[:,-1].to_numpy();
print("Number of samples in the Chocolate dataset: " + str(len(np_samples_choco)));
print("Number of features in a sample: " + str(len(np_samples_choco[random.randint(0, len(np_samples_choco)-1)])));
print("Number of Labels in the Chocolate dataset: " + str(len(np_labels_choco)));
print("Unique Labels: " + str(np.unique(np_labels_choco)));

Number of samples in the Chocolate dataset: 1885
Number of features in a sample: 13
Number of Labels in the Chocolate dataset: 1885
Unique Labels: ['non-user' 'user']


In [12]:
# Splitting
choco_X_train, choco_X_test, choco_y_train, choco_y_test = train_test_split(
    np_samples_choco,
    np_labels_choco,
    test_size=1/3,
    random_state=random.randint(0, 2**32 - 1)
);

In [13]:
# Verifying the Data
print("Number of Samples in the training set: " + str(len(choco_X_train)));
print("Number of Labels in the training set: " + str(len(choco_y_train)));
print("Unique Labels in the training set: " + str(np.unique(choco_y_train)));

print("Number of Samples in the test set: " + str(len(choco_X_test)));
print("Number of Labels in the test set: " + str(len(choco_y_test)));
print("Unique Labels in the test set: " + str(np.unique(choco_y_test)));

Number of Samples in the training set: 1256
Number of Labels in the training set: 1256
Unique Labels in the training set: ['non-user' 'user']
Number of Samples in the test set: 629
Number of Labels in the test set: 629
Unique Labels in the test set: ['non-user' 'user']


In [14]:
# Verify whether the sizes of training set and the test set add up to the total size of the dataset
print(len(choco_X_train) + len(choco_X_test) == len(df_dataset_choco.to_numpy()));

True


Observations:
- We successfully split the dataset into a training set and a test set,
- with the same number of samples and labels in each set, 
-  and reserving all possible labels from the dataset correctly.
- Meanwhile, the splitting did not trim or remove any data records.

In [15]:
# Save the training and testing data for Chocolate dataset
joblib.dump(choco_X_train, constants.choco_dataset + '_train-set_samples.pkl');
joblib.dump(choco_y_train, constants.choco_dataset + '_train-set_labels.pkl');
joblib.dump(choco_X_test, constants.choco_dataset + '_test-set_samples.pkl');
joblib.dump(choco_y_test, constants.choco_dataset + '_test-set_labels.pkl');

#### Train each Model

In [16]:
# Set the training data into models
chocoModels.set_X_train(choco_X_train);
chocoModels.set_y_train(choco_y_train);
# Set the testing data into models
chocoModels.set_X_test(choco_X_test);
chocoModels.set_y_test(choco_y_test);

In [17]:
# Normalize data from training set and testing set respectively
chocoModels.normalize();

In [18]:
# Train all models
chocoModels.train();

Training all models
All models are completely trained.


After training the classifiers with the Chocolate dataset, some files were exported. Hence, we arrange the, in proper file organization.

In [19]:
organize();

Moved choc_model_decisionTree_posttrained.pkl to ./choc\posttrained
Moved choc_model_decisionTree_pretrained.pkl to ./choc\pretrained
Moved choc_model_Gradient_Boosting_posttrained.pkl to ./choc\posttrained
Moved choc_model_Gradient_Boosting_pretrained.pkl to ./choc\pretrained
Moved choc_model_KNN_posttrained.pkl to ./choc\posttrained
Moved choc_model_KNN_pretrained.pkl to ./choc\pretrained
Moved choc_model_MLP_posttrained.pkl to ./choc\posttrained
Moved choc_model_MLP_pretrained.pkl to ./choc\pretrained
Moved choc_model_randomForest_posttrained.pkl to ./choc\posttrained
Moved choc_model_randomForest_pretrained.pkl to ./choc\pretrained
Moved choc_model_SVC_RBF_posttrained.pkl to ./choc\posttrained
Moved choc_model_SVC_RBF_pretrained.pkl to ./choc\pretrained
Moved choc_test-set_labels.pkl to ./choc\Test Set
Moved choc_test-set_samples.pkl to ./choc\Test Set
Moved choc_train-set_labels.pkl to ./choc\Training Set
Moved choc_train-set_samples.pkl to ./choc\Training Set
Organization complet